src/main/java/edu/cornell/kfs/module/ar/document/web/struts/CuContractsGrantsInvoiceDocumentAction.java

```diff
@@ -41,6 +56,115 @@ public ActionForward prorateBill(ActionMapping mapping, ActionForm form, HttpSer
+                }
+
+                Pair<Date, Date> awardDateRange = parseDateRange(contractsGrantsInvoiceDocument.getInvoiceGeneralDetail().getAwardDateRange());
+                Date awardStartDate = awardDateRange.getKey();
+                if (dateTimeService.dateDiff(billingPeriodStartDate, awardStartDate, false) >= 1) {
+                    warningMessages.add(configurationService.getPropertyValueAsString(CUKFSKeyConstants.WARNING_CINV_BILLING_PERIOD_START_DATE_BEFORE_AWARD_START_DATE));
+                }
+            } catch(ParseException ex) {
+                LOG.error("getContractsGrantsInvoiceDocumentWarningMessage: " + ex.getMessage());
+                warningMessages.add("ParseException occurred while parsing the billing period. Do you want to Proceed?[br]" + ex.getMessage());
+            }
+        }
+
+        String ret = CollectionUtils.isEmpty(warningMessages) ? null : StringUtils.join(warningMessages, "[br]");
+        return ret;
+    }
+
+    protected ActionForward promptForFinalBillConfirmation(ActionMapping mapping, ActionForm form, HttpServletRequest request, HttpServletResponse response, String caller, String warningMessage, ContractsGrantsInvoiceDocument contractsGrantsInvoiceDocument) throws Exception {
+        ActionForward forward = null;
+
+        Object question = request.getParameter(KFSConstants.QUESTION_INST_ATTRIBUTE_NAME);
+        if (question == null) {
+            return performQuestionWithoutInput(mapping, form, request, response, CuArConstants.CINV_FINAL_BILL_INDICATOR_CONFIRMATION_QUESTION, warningMessage, KFSConstants.CONFIRMATION_QUESTION, caller, StringUtils.EMPTY);
+        }
+
+        Object buttonClicked = request.getParameter(KFSConstants.QUESTION_CLICKED_BUTTON);
+        if (CuArConstants.CINV_FINAL_BILL_INDICATOR_CONFIRMATION_QUESTION.equals(question)) {
+            if (ConfirmationQuestion.NO.equals(buttonClicked)) {
+                forward = mapping.findForward(KFSConstants.MAPPING_BASIC);
+            }
+            else if (ConfirmationQuestion.YES.equals(buttonClicked) && contractsGrantsInvoiceDocument.getInvoiceGeneralDetail().isFinalBillIndicator()) {
+                try {
+                    Pair<Date, Date> billingPeriod = parseDateRange(contractsGrantsInvoiceDocument.getInvoiceGeneralDetail().getBillingPeriod());
+                    Date billingEndDate = billingPeriod.getValue();
+                    DateTimeService dateTimeService = SpringContext.getBean(DateTimeService.class);
+                    if (dateTimeService.dateDiff(billingEndDate, contractsGrantsInvoiceDocument.getInvoiceGeneralDetail().getLastBilledDate(), false) != 0) {
+                        contractsGrantsInvoiceDocument.getInvoiceGeneralDetail().setLastBilledDate(billingEndDate);
+                    }
+                }
+                catch(Exception ex) {
+                    LOG.error("promptForFinalBillConfirmation setting last billed date to end date: " + ex.getMessage());
```

Comment:
Should execution of the action still continue if an error happens here, or should the document be returned to the user with an explanation of the error?

Reply:
I think that it would be a better user experience practice to stop and explain the error. Should I use GlobalVariables.getMessageMap().putError?

Reply:
Yes, using GlobalVariables.getMessageMap().putError would be the best option here. Just make sure you use the KFS version of GlobalVariables and not the Rice one.

Reply:
I agree that this section needs to add an error to GlobalVariables and return to the user. Please update this area accordingly.